### DISCUSSION TITLE
api post calls getting typed as void

### DISCUSSION THREAD

**Discussion by gavinthomas-valtech** [orval-labs/orval#784] [Q&A]
Wondering why my api post requests are typed as `<void>` 

What would be in the yaml file that would define this?

how do i get the api team to adjust the yaml file so they are typed?

all get requests are fine

``` 
return apiClient<void>({
    url: `/blah/
      method: 'post',
    ...
    ```

**Comment by gavinthomas-valtech** [comment]
I guess in this case because there is no 201 response specified it's creating <void> when in fact i want it to use the values passed in

**Comment by gavinthomas-valtech** [comment]
Still can't work this out. The swagger/openapi doc has a schema for a 400

![image](https://user-images.githubusercontent.com/118996919/226842896-eef09f81-aa99-406c-bb97-c92315f9760f.png)

but this is always getting typed as `<void>`

eg

```
xport const updateAnonymousElectorApplicationPhotoStatus = (
  Id: string,
  applicationId: string,
  photoId: string,
  updateApplicationPhotoStatusBody: UpdateApplicationPhotoStatusBody
) => {
  return apiClient<void>({
    url: `/${photoId}`,
    method: 'post',
    headers: { 'Content-Type': 'application/json' },
    data: updateApplicationPhotoStatusBody,
  });
};
```

**Reply by anymaniax** [reply]
For the client axios, orval only takes the success status that's probably why. Since if you receive a 400 you will be in the error flow. You  will need to cache it and with axios there is no way to type the response error directly. The model of the error is still generated so you can still use it somewhere if needed

**Reply by gavinthomas-valtech** [reply]
Can it only ever type `<TData>` . How do we get it to type <TData, TError> or <SuccessResponse,ErrorResponse>

**Reply by anymaniax** [reply]
You need to handle it yourself. Axios and promise don't provide that directly. You could probably accomplish that using a [mutator](https://orval.dev/reference/configuration/output#mutator). I gonna try to do an example for you today

**Reply by gavinthomas-valtech** [reply]
cool. Thanks for clarifying 👍

**Reply by gavinthomas-valtech** [reply]
>You could probably accomplish that using a [mutator](https://orval.dev/reference/configuration/output#mutator). I gonna try to do an example for you today

amazing! thanks. Have had great feedback in the team using this client. Looking to do a demo to others in the agency (global agency)

**Reply by anymaniax** [reply]
I just checked and currently, orval doesn't provide the error type to the mutator could be an improvement to be added to be able to do that. But if you are using something like `react-query`. It should work properly with the error type too.

**Reply by gavinthomas-valtech** [reply]
Yes using react query but not as a setting in orval. Creating custom hooks that consume axios functions with a custom mutator. Mainly so we can keep adding to.

This was the post that put me onto this library https://profy.dev/article/react-openapi-typescript

**Reply by gavinthomas-valtech** [reply]
@anymaniax how did you get on with custom mutator client? Really interested to see this and how it will help

**Reply by ellemedit** [reply]
this doesn't help generate types for 400 status code responses.
